### PR TITLE
feat(model): Add Huggingface trainer

### LINKED ISSUE #310: Continue Pretraining Llama7B with Huggingface trainer
Information in requirement:
https://docs.google.com/document/d/1DQ5a56gFv2ZzRKn9PliNlOc-d-CI5vGAUUsdPihX4OI/edit#heading=h.5bb1u3i0p4rh

Additional Steps:
1 Understand Lanta Multinode training guidelines first https://app.gitbook.com/o/ygzlt6vZbi4mM0I2X5ko/s/rXqu9ENRkozaiYy0LTZK/lanta/multinode-training
2. Try to integrate multinode training with the Huggingface Trainer script
3. Integrate our own Huggingface dataset (V5_555) into Training code
4. Run all [Checklist](https://app.gitbook.com/o/ygzlt6vZbi4mM0I2X5ko/s/rXqu9ENRkozaiYy0LTZK/testing/model-training-checklist)
5. Notify @kwan or @boat or @new before running full training
6. Run Training Note: This step is **blocked by** [Task:Integrate Data Pipeline into OSCAR Colossal](https://github.com/orgs/OpenThaiGPT/projects/10/views/1?filterQuery=sprint%3A%22Sprint+3%22+&pane=issue&itemId=39481051)

**Important Note:** No need for fancy config script, or codebase. Just one train.py (200-400 lines) is enough for this task. Recommend modifying Standford Alpaca training code from **Multinode training guidelines**

Recommended People to do the task (Pick 1)
1. @New
2. @Boss
3. @Boat
4. @Bank
5. @Tae

### PR DESCRIPTION
**Pull request by boss-chanon**
## Why this PR
Huggingface trainer efficient more than old pipeline
@new5558 approve have more than 200 lines changes

## Changes
- add hugginface trainer scripts

## Related Issues
Close #310

## Checklist
- [x] Resume training from checkpoint
- [x] Model exporting tests
- [x] Fitting on small dataset
- [x] Muti GPU/Node testing

### CODE REVIEW

**Comment by new5558** [comment]
I think we only need deepspeed code for now. please remove FSDP related submission code.

File: src/model/scripts/hf_trainer/sbatch submit_deepspeed.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+# apptainer exec  --bind  /project/lt200056-opgpth/stanford_alpaca:/home/nbuppodo --writable pytorch_image_fixed pip install git+https://github.com/huggingface/accelerate
+
+START=`date`
+starttime=$(date +%s)
+
+
+export WANDB_MODE="offline"
+
+torchrun --nproc_per_node=4 --master_port=3000 train.py \
+    --model_name_or_path /project/lt200056-opgpth/boss/stanford_alpaca/llama_2_7b_fixed_resized \
```
Comment (new5558) [inline src/model/scripts/hf_trainer/sbatch submit_deepspeed.sh:36]: Can you make sure this model is using thai tokenizer (The merged vocab one)

File: src/model/scripts/hf_trainer/sbatch submit_deepspeed.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    --tokenizer_name_or_path /project/lt200056-opgpth/boss/stanford_alpaca/llama_2_7b_fixed_resized \
```
Comment (new5558) [inline src/model/scripts/hf_trainer/sbatch submit_deepspeed.sh:37]: Same here, Can you make sure this model is using thai tokenizer

File: src/model/scripts/hf_trainer/smultinode_deepspeed.sh [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+
+
+# source /fsx/dalle2/.dalle_env_38/bin/activate
+# echo python3 version = `python3 --version`
+# python -c "import torch"
+
+
+accelerate launch --num_processes $(( 4 * $COUNT_NODE )) --num_machines $COUNT_NODE --multi_gpu --mixed_precision fp16 --machine_rank $SLURM_PROCID --main_process_ip $MASTER_ADDR --main_process_port $MASTER_PORT  train.py \
+    --model_name_or_path /project/lt200056-opgpth/boss/stanford_alpaca/llama_2_7b_fixed_resized \
```
Comment (new5558) [inline src/model/scripts/hf_trainer/smultinode_deepspeed.sh:43]: Can you make sure this model is using thai tokenizer

File: src/model/scripts/hf_trainer/smultinode_deepspeed.sh [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    --tokenizer_name_or_path /project/lt200056-opgpth/boss/stanford_alpaca/llama_2_7b_fixed_resized \
```
Comment (new5558) [inline src/model/scripts/hf_trainer/smultinode_deepspeed.sh:44]: Can you make sure this model is using thai tokenizer

File: src/model/scripts/hf_trainer/smultinode_deepspeed.sh [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    --data_path /scratch/lt200056-opgpth/hf_v5_token_llama_2_256 \
```
Comment (new5558) [inline src/model/scripts/hf_trainer/smultinode_deepspeed.sh:45]: I think this should be 1024 sequence length according to the requirement on sprint card

**Comment by new5558** [comment]
image
image

Please adhere to model training requirement

**Comment by new5558** [comment]
Please run through (Model Training Checklist)[https://app.gitbook.com/o/ygzlt6vZbi4mM0I2X5ko/s/rXqu9ENRkozaiYy0LTZK/testing/model-training-checklist] especially resume from checkpoint

File: src/model/scripts/hf_trainer/sbatch submit_deepspeed.sh (hunk 1 above)
Comment (boss-chanon) [reply src/model/scripts/hf_trainer/sbatch submit_deepspeed.sh:36]: already check

File: src/model/scripts/hf_trainer/sbatch submit_deepspeed.sh (hunk 2 above)
Comment (boss-chanon) [reply src/model/scripts/hf_trainer/sbatch submit_deepspeed.sh:37]: already check

File: src/model/scripts/hf_trainer/smultinode_deepspeed.sh (hunk 4 above)
Comment (boss-chanon) [reply src/model/scripts/hf_trainer/smultinode_deepspeed.sh:44]: already check

File: src/model/scripts/hf_trainer/smultinode_deepspeed.sh (hunk 3 above)
Comment (boss-chanon) [reply src/model/scripts/hf_trainer/smultinode_deepspeed.sh:43]: already check

File: src/model/scripts/hf_trainer/smultinode_deepspeed.sh (hunk 5 above)
Comment (boss-chanon) [reply src/model/scripts/hf_trainer/smultinode_deepspeed.sh]: done

**Review by boat1603 — APPROVED** [review]

**Review by new5558 — APPROVED** [review]

**Pull request merged**